src/formats/pnm.c

```diff
@@ -0,0 +1,287 @@
+// files, but this is again not really supported. Thus, we don't support either
+// feature for consistency with other parsers
+
+// PNM file types
+#define PNM_PBM 0
+#define PNM_PGM 1
+#define PNM_PPM 2
+
+// A file-like abstraction for cleaner number parsing
+struct pnm_iter {
+    const uint8_t* pos;
+    const uint8_t* end;
+};
+
+// Error conditions
+#define PNM_EEOF -1
+#define PNM_ERNG -2
+#define PNM_EFMT -3
+#define PNM_EOVF -4
+
+static const char* pnm_strerror(int err)
+{
+    switch (err) {
+        case PNM_EEOF:
+            return "unexpected end of image";
+        case PNM_ERNG:
+            return "integer too large";
+        case PNM_EFMT:
+            return "digit expected";
+        case PNM_EOVF:
+            return "pixel value above maxval";
+        default:
+            return "unknown error";
+    }
+}
+
+// Read an integer, ignoring leading whitespace and comments
+static int pnm_readint(struct pnm_iter* it)
+{
+    for (; it->pos != it->end; ++it->pos) {
+        if (*it->pos == '#') {
+            while (it->pos != it->end && *it->pos != '\n' && *it->pos != '\r')
+                ++it->pos;
+        } else if (*it->pos != ' ' && *it->pos != '\t' && *it->pos != '\n' &&
+                   *it->pos != '\r') {
+            // XXX isspace is still on the table, but every parser I read
+            // handled this themselves
+            break;
+        }
+    }
+    if (it->pos == it->end)
+        return PNM_EEOF;
+    if (*it->pos < '0' || *it->pos > '9')
+        return PNM_EFMT;
+    int val = 0;
+    // XXX strtol is still on the table, but to do so we'd need to make a first
+    // pass to copy to a buffer, so it seems easier to stick with the majority
+    // and do this ourselves as well
+    do {
+        uint8_t d = *it->pos - '0';
+        // XXX the standard doesn't mention how large integers may be; most of
+        // the parsers work with signed integers only
+        if (val > INT_MAX / 10)
+            return PNM_ERNG;
+        val *= 10;
+        if (val > INT_MAX - d)
+            return PNM_ERNG;
+        val += d;
+        ++it->pos;
+    } while (it->pos != it->end && *it->pos >= '0' && *it->pos <= '9');
+    return val;
+}
+
+// Decode a plain/ASCII PNM file; returns 0 on success, an error code otherwise
+static int decode_plain(struct image_frame* f, struct pnm_iter* it, int type,
+                        int maxval)
+{
+    for (size_t y = 0; y < f->height; ++y) {
+        argb_t* dst = f->data + y * f->width;
+        for (size_t x = 0; x < f->width; ++x) {
+            argb_t pix = ARGB_SET_A(0xff);
+            if (type == PNM_PBM) {
+                int bit = pnm_readint(it);
+                if (bit < 0)
+                    return bit;
+                if (bit > 1)
+                    return PNM_EOVF;
+                pix |= bit - 1;
+            } else if (type == PNM_PGM) {
+                int v = pnm_readint(it);
```

Comment:
Can be `const`.
Please, add the `const` qualifier to read-only variables in other places.